include/utilities/dependency_graph.py

```diff
@@ -0,0 +1,101 @@
+from collections import OrderedDict
+from enum import Enum
+from typing import TypeVar, Generic
+
+
+T = TypeVar('T')
+
+
+class NodeStatus(Enum):
+    NotVisited = 0
+    Visiting = 1
+    Visited = 2
+
+
+class CircularDependencyException(Exception, Generic[T]):
+    _chain: list[T]
+
+    def __init__(self, *chain: T):
+        self._chain = list(chain)
+
+    @property
+    def chain(self):
+        return self._chain
+
+
+class DependencyGraph(Generic[T]):
+    _graph: dict[T, list[T]]
+
+    def __init__(self):
+        self._graph = {}
+
+    def get_dependency_order(self) -> list[list[T]]:
+        visit_status: dict[T, NodeStatus] = {
+            item: NodeStatus.NotVisited for item in self._graph
+        }
+        _levels: dict[T, int] = {}
+        _error_nodes: set[T] = set()
+
+        def _get_max_dependency_level(dependant: T) -> int:
+
+            if visit_status[dependant] == NodeStatus.Visited:
+                return _levels[dependant]
+            visit_status[dependant] = NodeStatus.Visiting
+
+            max_dependency_level = 0
+            max_dependency: T | None = None
+
+            for dependency in self._graph[dependant]:
+                # if dependency is dependant:
+                #     raise CircularDependencyException[T](dependency)
```

Comment:
:D

Reply:
I took that as it is from Z#

Reply:
And I'll not change that until I see real usage of the dependency graph.